Support for SSE events via openapigen

timbms opened the pull request.

timbms committed:
Include tag 'events' in openapi generation

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

timbms committed:
Correcting openapi.json for Content-Type sent by getEvents

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

timbms commented:
The openAPI description is missing the return type for getEvents.

digitaldan commented on OpenHABCore/Sources/OpenHABCore/GeneratedSources/openapi/Client.swift:
```diff
@@ -2757,6 +2757,171 @@ public struct Client: APIProtocol {
+                    )
+                    switch chosenContentType {
+                    case "text/event-stream":
+                        body = try converter.getResponseBodyAsBinary(
```
Since this is an event stream and the data will will be read forever (until the connection is closed), will getResponseBodyAsBinary work here?

timbms replied:
This is the generated code from openapi. I would assume it works. I pushed it further and created a helper function to obtain an AsyncSequence that should contain the information in data parsed into OpenHABEvents that I derived from heuristics

digitaldan commented on OpenHABCore/Sources/OpenHABCore/GeneratedSources/openapi/Client.swift:
```diff
@@ -2757,6 +2757,171 @@ public struct Client: APIProtocol {
+            },
+            deserializer: { response, responseBody in
+                switch response.status.code {
+                case 200:
```
This is an event stream , where  the initial `ready` message returns the UUID used in `/rest/events/states/{uuid}` to add items to listen to. so i don't this this deserializer is going to work? 

<img width="917" height="110" alt="image" src="https://github.com/user-attachments/assets/07c9b2a3-510e-45b4-8700-a91a29e15c4e" />

I'll maybe poke around and see how to best handle SSE endpoints in the swift openapi stuff, unless @timbms you have some ideas here?

timbms replied:
I don't have experience with SSE yet
Still a lot of discovery. 
Also in how to complete the gaps in openHAB's openapi.json

digitaldan commented on OpenHABCore/Sources/OpenHABCore/openapi/openapi.json:
```diff
@@ -5890,7 +5890,11 @@
         "responses": {
           "200": {
-            "description": "OK"
+            "description": "OK",
+            "content": {
+              "text/event-stream": {},
+              "application/json": {}
+            }
```
Sorry if i'm explaining the obvious (I'm learning here as i go), should these content types also be added to `/events/state` as well to enable SSE there ?

timbms replied:
After looking at the output received from  `curl http://192.168.2.10:8080/rest/events -i ` :
```      
HTTP/1.1 200 OK
Content-Type: text/event-stream
X-Accel-Buffering: no
Content-Encoding: identity
Transfer-Encoding: chunked
Server: Jetty(9.4.57.v20241219)

event: message
data: {"topic":"openhab/items/SprudelsteinShellyPlug5shellyplugs9d903d1921682114_Watt/state","payload":"{\"type\":\"Quantity\",\"value\":\"23.16 W\"}","type":"ItemStateEvent"}

event: message
data: {"topic":"openhab/items/SprudelsteinShellyPlug5shellyplugs9d903d1921682114_TotalEnergy/state","payload":"{\"type\":\"Quantity\",\"value\":\"0.257 kWh\"}","type":"ItemStateEvent"}

event: message
data: {"topic":"openhab/items/SprudelsteinShellyPlug5shellyplugs9d903d1921682114_Watt/stateupdated","payload":"{\"type\":\"Quantity\",\"value\":\"23.16 W\",\"lastStateUpdate\":\"2025-07-30T15:40:52.567453494+02:00[Europe/Berlin]\"}","type":"ItemStateUpdatedEvent"}
```
with field names `event` and `data` I think we can drop  "application/json"

digitaldan replied:
Yeah, i have modified the JSON and its just using that response type.

Sorry, i should have mentioned so we don't duplicate work..... i'm working on your branch now, i'll go ahead and merge some code when i get it working, i think i'm pretty close.   The PR was a great place to start, i see if i can finish it since i have a vested interest in the end feature.

timbms replied:
Go ahead. I just wanted to pass the knowledge I have. 
I am curious to see what you will make out of it.

timbms committed:
Removed content application/json for getEvents and regenerated Client.swift and Types.swift
Introduced the struct OpenHABEvent to decode getEvents data. Should be included in openapi.json
Introduced helper openHABEvents(topics: ) to get an AsyncSequence

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

timbms committed:
DateTime parsing pushed to OpenAPIService initializer
Removed updateForLongPolling and pollDataForSitemap(subscriptionId) that was not used

Signed-off-by: Tim Mueller-Seydlitz <timbms@gmail.com>

digitaldan committed:
Adds a complete working SSE command item impl

Signed-off-by: Dan Cunningham <dan@digitaldan.com>

digitaldan commented:
Ok, so SSE is working through the API generation stuff, i also included a fully working command item feature, where the same commands that can be sent via a push notification, can now be sent via a String command item (which is saved with the home profile).   There's also support for showing/hiding the screensaver, which i will come back to a bit later and formalize some naming patterns there.  We should also be able to do things like enable/disable idle timer, set screen brightness, take a screenshot or use the camera to take a pic.......

digitaldan commented:
Just FYI something i did wipes out the previous openHAB settings on the device.  I'm not sure if an ID changed  in the project file thats used to identify preferences, or if adding a new preference did something....... i need to look into that.  I think going forward it save and load them just fine.

digitaldan committed:
revert dev id

Signed-off-by: Dan Cunningham <dan@digitaldan.com>

digitaldan committed:
remove commented code

Signed-off-by: Dan Cunningham <dan@digitaldan.com>

digitaldan committed:
better screensaver commands

Signed-off-by: Dan Cunningham <dan@digitaldan.com>

digitaldan committed:
new device app options for notifications/sse

Signed-off-by: Dan Cunningham <dan@digitaldan.com>

digitaldan committed:
Adds TTS, finalizes device sytax, updates README

Signed-off-by: Dan Cunningham <dan@digitaldan.com>

digitaldan commented:
Ok, i think this is ready.   I went for some extra credit here and added text to speech (tts)  to the app ;-).  So now you can announce stuff from the device using customized voices!

digitaldan committed:
remove comment

Signed-off-by: Dan Cunningham <dan@digitaldan.com>

digitaldan committed:
Small cleanups

Signed-off-by: Dan Cunningham <dan@digitaldan.com>
